File: shared/web-sdk/src/audio-pre-roll-ring.ts
// ---------------------------------------------------------------------------
// AudioPreRollRing — capture-side pre-roll + hangover buffer.
//
// PROBLEM
// -------
// The client-side EchoGate drops mic frames whose RMS falls below a baseline
// threshold (room noise, between-word silence, very quiet onset/offset
// phonemes). When real speech begins, its leading 1-2 frames are often
// quieter than the threshold (vowel ramp-up, partial frame of silence
// before the syllable starts) and get dropped — the STT never sees them,
// so transcripts miss the head of the utterance. Symmetrically, trailing
// low-energy phonemes drop off the tail. Worse, mid-sentence dips below
// the threshold cause the gateway's silence-carrier to engage, which makes
// Silero VAD fire end-of-speech and Smart-Turn cut the turn — one utterance
// becomes several transcripts.
//
// DESIGN
// ------
// A small ring placed between the EchoGate decision and the WS send. Every
// captured frame is run through the gate as today, but instead of dropping
// rejected frames outright we buffer the last N of them (pre-roll). When the
// gate flips from reject → accept, the buffered pre-roll is flushed ahead
// of the accepted frame so STT sees the quiet onset. When the gate flips
// from accept → reject, we keep emitting for M more frames (hangover) so
// the quiet tail is preserved.
//
// State machine (pure, no clocks):
//
//     idle (last frame rejected, ring filling)
//       — accept → flush ring, emit frame, → active
//       — reject → push to ring (trim), emit nothing
//
//     active (last frame accepted, hangover counter armed)
//       — accept → emit frame, re-arm hangover, stay active
//       — reject → emit frame, decrement hangover; → idle when counter hits 0
//
// PLATFORM PORTING
// ----------------
// Pure TypeScript, zero browser deps. Same porting story as EchoGate.
// ---------------------------------------------------------------------------

export interface AudioPreRollRingConfig {
  /**
   * Number of below-threshold frames to retain ahead of the next accepted
   * frame. At ~80 ms per frame, 5 ≈ 400 ms of leading context — enough to
   * cover a slow vowel ramp-up or a quiet leading consonant.
   */
  readonly preRollFrames: number;
  /**
   * Number of below-threshold frames to keep emitting after an accepted
   * streak. Covers trailing low-energy phonemes and inter-word dips so
   * the gateway's silence carrier does not engage mid-sentence. At
   * ~80 ms per frame, 4 ≈ 320 ms of hangover.
   */
  readonly hangoverFrames: number;
}

export interface AudioPreRollRing {
  /**
   * Feed one captured frame with the EchoGate's accept/reject verdict.
   * Returns the frames that should be forwarded to STT in order — may be
   * empty (silence is being buffered) or contain multiple frames (the
   * pre-roll flush on a reject→accept transition).
   */
  push(frame: ArrayBuffer, accepted: boolean): readonly ArrayBuffer[];
  /** Drop the ring and reset to idle. Use on session boundaries. */
  reset(): void;
}

export function createAudioPreRollRing(config: AudioPreRollRingConfig): AudioPreRollRing {
  if (config.preRollFrames < 0) {
    throw new Error(`AudioPreRollRing: preRollFrames must be >= 0, got ${config.preRollFrames}`);
  }
  if (config.hangoverFrames < 0) {
    throw new Error(`AudioPreRollRing: hangoverFrames must be >= 0, got ${config.hangoverFrames}`);
  }

  const ring: ArrayBuffer[] = [];
  let active = false;
  let hangoverRemaining = 0;

  return {
    push(frame, accepted) {
      if (accepted) {
        if (!active) {
          const flush = ring.length > 0 ? ring.slice() : [];
          ring.length = 0;
          flush.push(frame);
          active = true;
          hangoverRemaining = config.hangoverFrames;
          return flush;
        }
        hangoverRemaining = config.hangoverFrames;
        return [frame];
      }

      if (active) {
        if (hangoverRemaining > 0) {
          hangoverRemaining--;
          if (hangoverRemaining === 0) active = false;
          return [frame];
        }
        active = false;
      }

      if (config.preRollFrames === 0) return [];
      ring.push(frame);
      while (ring.length > config.preRollFrames) ring.shift();
      return [];
    },
    reset() {
      ring.length = 0;
      active = false;
      hangoverRemaining = 0;
    },
  };
}

File: shared/mobile-sdk/src/commonMain/kotlin/io/sentient/mobilesdk/audio/AudioPreRollRing.kt
package io.sentient.mobilesdk.audio

// Retained for Slice 2 (downlink playback-window reuse); not wired into the current uplink.
// ---------------------------------------------------------------------------
// AudioPreRollRing — capture-side pre-roll + hangover buffer.
//
// Port of shared/web-sdk/src/audio-pre-roll-ring.ts.
//
// PROBLEM
// -------
// The client-side EchoGate drops mic frames whose RMS falls below a baseline
// threshold. When real speech begins its leading 1–2 frames are often quiet
// (vowel ramp-up, partial frame of silence before the syllable) and get
// dropped — STT misses the utterance head. Symmetrically, trailing low-energy
// phonemes drop off the tail, and mid-sentence dips engage the gateway's
// silence-carrier, which causes Silero VAD to fire end-of-speech early.
//
// DESIGN
// ------
// A small ring placed between the EchoGate decision and the upstream send.
// Every captured frame is run through the gate; instead of dropping rejected
// frames outright, the last N are buffered (pre-roll). On a reject→accept
// transition the buffer is flushed ahead of the accepted frame. On an
// accept→reject transition frames are kept emitting for M more frames
// (hangover) before going silent.
//
// STATE MACHINE (mirrors audio-pre-roll-ring.ts verbatim)
// --------------------------------------------------------
//   idle (last frame rejected, ring filling)
//     — accept → flush ring, emit frame → active
//     — reject → push to ring (trim), emit nothing
//
//   active (last frame accepted, hangover counter armed)
//     — accept → emit frame, re-arm hangover, stay active
//     — reject → emit frame, decrement hangover; → idle when counter hits 0
//              → if hangoverRemaining already 0 on entry: idle immediately
//
// BRANCH ORDER (verified against TS, line by line)
// --------------------------------------------------
// 1. accepted=true, !active  → onset: flush ring + frame, go active
// 2. accepted=true, active   → re-arm hangover, return [frame]
// 3. accepted=false, active, hangoverRemaining > 0
//      → decrement; if hits 0, go idle; return [frame]
// 4. accepted=false, active, hangoverRemaining == 0
//      → go idle (fall through to idle branch)
// 5. accepted=false, idle, preRollFrames == 0 → return empty
// 6. accepted=false, idle    → push to ring, trim to preRollFrames, return empty
//
// TS PARITY NOTE
// --------------
// The reference Kotlin impl in the task plan matches the TS exactly. No
// corrections were required — branch ordering and off-by-one (trim uses
// `> preRollFrames`, not `>=`) agree with the TS source.
// ---------------------------------------------------------------------------

/** Maximum frames to retain ahead of the next accepted frame (pre-roll size). */
private const val MIN_FRAMES = 0

/**
 * AudioPreRollRing gates mic uplink with a pre-roll buffer and a hangover tail.
 *
 * Generic over [T] so it works with any frame type (ByteArray, ShortArray, Int
 * in tests). Pure logic — no platform deps, no clocks, no logging.
 *
 * @param preRollFrames Number of below-threshold frames to retain before onset.
 *                      Must be >= 0.
 * @param hangoverFrames Number of below-threshold frames to keep emitting after
 *                       an accepted streak. Must be >= 0.
 */
class AudioPreRollRing<T>(
    private val preRollFrames: Int,
    private val hangoverFrames: Int,
) {
    init {
        require(preRollFrames >= MIN_FRAMES) {
            "AudioPreRollRing: preRollFrames must be >= 0, got $preRollFrames"
        }
        require(hangoverFrames >= MIN_FRAMES) {
            "AudioPreRollRing: hangoverFrames must be >= 0, got $hangoverFrames"
        }
    }

    private val ring = ArrayDeque<T>()
    private var active = false
    private var hangoverRemaining = 0

    // -----------------------------------------------------------------------
    // Public API
    // -----------------------------------------------------------------------

    /**
     * Feed one captured frame with the EchoGate's accept/reject verdict.
     *
     * Returns the frames that should be forwarded to STT in order — may be
     * empty (silence is being buffered) or contain multiple frames (the
     * pre-roll flush on a reject→accept transition).
     */
    fun push(frame: T, accepted: Boolean): List<T> {
        if (accepted) {
            return if (!active) {
                // Onset: flush pre-roll ring + accepted frame, go active.
                val flush = ring.toMutableList()
                ring.clear()
                flush.add(frame)
                active = true
                hangoverRemaining = hangoverFrames
                flush
            } else {
                // Already active: re-arm hangover, pass frame through.
                hangoverRemaining = hangoverFrames
                listOf(frame)
            }
        }

        if (active) {
            if (hangoverRemaining > 0) {
                hangoverRemaining--
                if (hangoverRemaining == 0) active = false
                return listOf(frame)
            }
            // Hangover exhausted — fall through to idle buffering.
            active = false
        }

        // Idle: buffer the frame in the pre-roll ring (no-op if size is 0).
        if (preRollFrames == 0) return emptyList()
        ring.addLast(frame)
        while (ring.size > preRollFrames) ring.removeFirst()
        return emptyList()
    }

    /** Drop the ring and reset to idle. Use on session boundaries. */
    fun reset() {
        ring.clear()
        active = false
        hangoverRemaining = 0
    }
}
